esmvaltool/preprocessor/_mask.py

```diff
@@ -300,63 +395,42 @@ def count_spells(data, threshold, axis, spell_length):
+def mask_window_threshold(cube, threshold_fraction, min_value, time_window):
     """
-    Find data counts in a time window
-
-    Function that returns a flat array containing
-    the number of data points within a time window `window_size'
-    per grid point that satify a condition
-    value > value_threshold.
-    It also returns statistical measures for the flat array
-    window_counts[0] = array
-    window_counts[1] = mean(array)
-    window_counts[2] = std(array)
-    window_counts[3] = percentile(array, pctile)
-    """
-    # Make an aggregator from the user function.
-    spell_count = Aggregator(
-        'spell_count', count_spells, units_func=lambda units: 1)
+    Find data counts in a time window.
 
-    # Calculate the statistic.
-    counts_windowed_cube = mycube.collapsed(
-        'time',
-        spell_count,
-        threshold=value_threshold,
-        spell_length=window_size)
-
-    # if one wants to print the whole array
-    # np.set_printoptions(threshold=np.nan)
-    r_p = counts_windowed_cube.data.flatten()
-    meanr = np.mean(r_p)
-    stdr = np.std(r_p)
-    prcr = np.percentile(r_p, pctile)
-    return r_p, meanr, stdr, prcr
-
-
-def mask_cube_counts(mycube, value_threshold, counts_threshold, window_size):
-    """Build the counts mask"""
-    # Make an aggregator from the user function.
-    spell_count = Aggregator(
-        'spell_count', count_spells, units_func=lambda units: 1)
+    Function that returns a masked data cube. The masking is done on
+    the number of valid data points within a time window `window_size'
+    per grid point that satify a simple value thresholding condition
+    value > value_threshold. This is the equivalent of mask_fillvalues
+    but for single data products (single iris cubes).
 
-    # Calculate the statistic.
-    counts_windowed_cube = mycube.collapsed(
-        'time',
-        spell_count,
-        threshold=value_threshold,
-        spell_length=window_size)
-
-    mask = counts_windowed_cube.data >= counts_threshold
-    mask.astype(np.int)
-    # preserving the original cube metadata
-    dummyar = np.ones(mycube.data.shape, dtype=mycube.data.dtype)
-    newmask = dummyar * mask
-    newmask[newmask == 0] = 1e+20  # np.nan
-    masked_cube = mycube.copy()
-    # masked_cube.data = masked_cube.data * newmask
-    masked_cube.data = newmask
-    return counts_windowed_cube, newmask, masked_cube
+    Parameters
+    ----------
+
+    * cube (iris.Cube.cube instance):
+        data cube to be masked.
+
+    * threshold_fraction (float, between 0 and 1):
+        fractional threshold of missing data points.
+
+    * min_value (float):
+        minumum value threshold.
+
+    * time_window (float):
+        time window to compute missing data counts.
+
+    Returns
+    -------
+    Masked iris cube.
+
+    """
+    mask = _get_fillvalues_mask(cube, threshold_fraction, min_value,
+                                time_window)
+    if not np.ma.isMaskedArray(cube.data):
```

Comment:
Maybe try something like `cube.data = da.ma.masked_array(cube.core_data())` to ensure you always have a masked array?